core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java

```diff
@@ -0,0 +1,369 @@
+  private Future<?> mBackupProgressFuture;
+  /** Scheduled future for timing out various backup stages. */
+  private ScheduledFuture<?> mBackupTimeoutTask;
+
+  /**
+   * Creates a new backup worker.
+   *
+   * @param masterContext the master context
+   */
+  public BackupWorkerRole(CoreMasterContext masterContext) {
+    super(masterContext);
+    // Read properties.
+    mBackupHeartbeatIntervalMs =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_HEARTBEAT_INTERVAL);
+    mLeaderConnectionIntervalMin =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_CONNECT_INTERVAL_MIN);
+    mLeaderConnectionIntervalMax =
+        ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_CONNECT_INTERVAL_MAX);
+  }
+
+  @Override
+  public void start() {
+    LOG.info("Starting backup-worker role.");
+    // Submit a task to establish and maintain connection with the leader.
+    mExecutorService.submit(() -> {
+      establishConnectionToLeader();
+    });
+  }
+
+  @Override
+  public void stop() throws IOException {
+    LOG.info("Stopping backup-worker role.");
+    // Cancel suspend timeout.
+    if (mBackupTimeoutTask != null && !mBackupTimeoutTask.isDone()) {
+      mBackupTimeoutTask.cancel(true);
+    }
+    // Cancel ongoing backup task.
+    if (mBackupFuture != null) {
+      mBackupFuture.cancel(true);
+    }
+    // Cancel heartbeat task.
+    if (mBackupProgressFuture != null) {
+      mBackupProgressFuture.cancel(true);
+    }
+    // Close leader close listener.
+    // This will ensure, connection won't be re-established when closed during stop.
+    if (mLeaderConnectionCloseListener != null) {
+      mLeaderConnectionCloseListener.close();
+    }
+    // Close the connection with the leader.
+    if (mLeaderConnection != null) {
+      mLeaderConnection.close();
+      mLeaderConnection = null;
+    }
+    // Stopping the base after because closing connection uses the base executor.
+    super.stop();
+  }
+
+  @Override
+  public Map<ServiceType, GrpcService> getRoleServices() {
+    return Collections.emptyMap();
+  }
+
+  @Override
+  public BackupStatus backup(BackupPRequest request) throws AlluxioException {
+    throw new IllegalStateException("Backup-worker role can't serve RPCs");
+  }
+
+  @Override
+  public BackupStatus getBackupStatus() throws AlluxioException {
+    throw new IllegalStateException("Backup-worker role can't serve RPCs");
+  }
+
+  /**
+   * Handler for suspend message. It's used in secondary master.
+   */
+  private void handleSuspendJournalsMessage(BackupSuspendMessage suspendMsg) {
+    LOG.info("Received suspend message: {}", suspendMsg.toString());
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    try {
+      mJournalSystem.suspend();
+      LOG.info("Suspended journals for backup.");
+    } catch (IOException e) {
+      String failMessage = "Failed to suspended journals for backup.";
+      LOG.error(failMessage, e);
```

Comment:
Do we need to log an error? We are already throwing an exception.

Reply:
Exceptions during message handling will show up only on the receiver side. This is to ease correlating.